modules/broker-core/src/main/java/org/wso2/broker/core/util/MessageTracer.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package org.wso2.broker.core.util;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.broker.core.Consumer;
+import org.wso2.broker.core.Message;
+import org.wso2.broker.core.Metadata;
+import org.wso2.broker.core.QueueHandler;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Trace messages flowing through the broker.
+ */
+public final class MessageTracer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MessageTracer.class);
+
+    public static final String NO_ROUTES = "No routes for message. Message dropped!";
+    public static final String UNKNOWN_EXCHANGE = "Unknown exchange. Message dropped!";
+    public static final String PUBLISH_FAILURE = "Failed to enqueue. Message dropped!";
+    public static final String PUBLISH_SUCCESSFUL = "Successfully enqueued message to queue.";
+    public static final String RETRIEVE_FOR_REDELIVERY = "Dequeue message for re-delivery.";
+    public static final String RETRIEVE_FOR_DELIVERY = "Dequeue message for delivery.";
+    public static final String REQUEUE = "Requeue message.";
+    public static final String ACKNOWLEDGE = "Acknowledge message.";
+    public static final String DELIVER = "Deliver message to transport consumer.";
+
+    public static void trace(Message message, String description) {
+        if (LOGGER.isTraceEnabled() && Objects.nonNull(message)) {
+            trace(message.getMetadata(), description);
+        }
+    }
+
+    public static void trace(Metadata metadata, String description) {
```

Comment:
Should this not be a private method? Are we calling the trace method passing metadata, outside the class?

Reply:
This is intentionally made public. We are calling this from outside the class.